design-proposals/amd-sev.md

```diff
@@ -0,0 +1,180 @@
+<domainCapabilities>
+...
+  <features>
+    ...
+    <sev supported='yes'>
+      <cbitpos>47</cbitpos>
+      <reducedPhysBits>1</reducedPhysBits>
+    </sev>
+    ...
+  </features>
+</domainCapabilities>
+```
+
+The `policy` parameter allows setting the guest policy flags. Additionally the optional `dhCert` and `session` provide the guest owners base64 encoded `DH` (Diffie-Hellman) key and the guest owners base64 encoded `launch blob` respectively.
+
+There are also several prerequisites to run SEV guests:
+
+- All virtio devices need to be configured with the `iommu='on'` attribute in their `<driver>` configuration
+- All memory regions used by the VM must be `locked` for Direct Memory Access (DMA) and to prevent swapping
+
+## [Attestation](https://github.com/AMDESE/sev-tool#proposed-provisioning-steps)
+
+### Certificate chain validation
+
+In order to ensure that the platform is not compromised the guest owner first needs to validate the certificate chain provided by the platform owner. On a system where KubeVirt is running, the privileged `virt-handler` pod can export the required certificates in advance to a `kbevirt-private` directory. Whenever a new VM requiring SEV is launched, the directory can be mounted to the `virt-launcher` pod in `read-only` mode. A user can then invoke the `virtctl` tool to download the certificate chain for validation (for that an additional command will be needed, e.g. `virtctl fetchcertchain ...`).
+
+### Validating the measurement
+
+The attestation for the basic SEV is performed before a VM is booted. It requers a user to provide the launch blob and the certificate (via `session` and `dhCert` paramters in the libvirt domain xml). A VM is launched in paused state thus allowing to retrieve and validate the measurement. Once it is verified, user can inject the secret into the VM and resume the execution.
+
+To support the attestation in KubeVirt `virtctl` can be extended to accept additional commands: `launchmeasure` and `injectsecret`. The commands will eventually end up with calling the corresponding libvirt API:
+
+- [virDomainGetLaunchSecurityInfo](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainGetLaunchSecurityInfo)
+
+    The call returns the launch measurement for a SEV guest.
+
+- [sev-inject-launch-secret](https://listman.redhat.com/archives/libvir-list/2021-May/msg00196.html)
+
+    The API is still missing.
+
+## Known limitations
+
+- SEV is only going to work with OVMF (UEFI)
+- SEV-encrypted VMs cannot contain directly-accessible host devices (that is, PCI passthrough)
+- Live Migration is not supported with SEV
+- The ammount of running SEV VMs on a node is limited
+- libvirt currently does not provide an API to inject a secret in a running VM
+
+Some of the limitations will probably be removed in the future.
+
+## API Examples
+
+**Requsting SEV feature in the VMI spec**
+
+```yaml
+spec:
+  domain:
+    launchSecurity:
+      sev:
+        policy:
+          - EncryptedState
```

Comment:
Based on your verification process which you describe, it sounds to me like we could however involve virtctl to do the heavy lifting of the whole verification process while still providing clear REST apis which also virtcl uses so that people could come up with their own flows entirely. Is that correct?

Reply:
Right, I though I had mentioned it somewhere in the doc that the idea was to have virtctl commands backed by corresponding API. So that virtctl can be used as a 'client' to interact with VMs during attestation (and maybe to facilitate the process a bit). At the same time the APIs will provide some space for further improvements.

Reply:
> Right, I though I had mentioned it somewhere in the doc that the idea was to have virtctl commands backed by corresponding API. So that virtctl can be used as a 'client' to interact with VMs during attestation (and maybe to facilitate the process a bit). At the same time the APIs will provide some space for further improvements.


What confused me is this section:

```
On a node where KubeVirt is running, the privileged `virt-handler` pod can export the required assets in advance to a `kubevirt-private` directory. Whenever a new VM requiring SEV is launched, the directory can be mounted to the `virt-launcher` pod in `read-only` mode. A user can then invoke the `virtctl` tool to download the certificate chain for validation (for that an additional command will be needed, e.g. `virtctl fetchcertchain ...`).
```

I somehow had the impression that here `virtctl` would do some strange things to get the certificates. maybe just add there that `virtctl` will query `virt-handler` to get the certificate and it works for me.